File: js/scheduler.js
/**
 * Deterministic roster allocation.
 *
 * A direct port of the original Python scheduler, kept behaviour-identical:
 * given the same inputs it always produces the same roster, every constraint it
 * is given is actually enforced, and anything it cannot satisfy is reported as
 * an explicit note rather than quietly fudged.
 *
 * The algorithm is a greedy fill with a fairness ranking. Days are processed in
 * order; within a day, shifts are filled in the organisation's own shift order.
 * For each vacancy the eligible pool is ranked by
 *
 *   1. fewest shifts worked so far in this roster   (spread the load)
 *   2. longest since last worked                    (spread the days out)
 *   3. name                                         (a stable, arbitrary tiebreak)
 *
 * Greedy rather than an exhaustive search because rosters are small, the result
 * must be explainable to whoever reads it ("Alice got Saturday because she'd
 * worked the fewest shifts"), and a solver that silently reshuffles everything
 * on a one-person change is worse to live with than one that does the obvious
 * thing.
 *
 * Dates are handled throughout as "YYYY-MM-DD" strings converted to whole-day
 * integers in UTC. Doing arithmetic on local-time Date objects would shift days
 * across daylight-saving boundaries, which for a roster is a real bug.
 */

/** Sentinel "has never worked" value; any real day sorts after it. */
const NEVER = -Infinity;

/** Convert a "YYYY-MM-DD" string to a whole number of days since the epoch. */
export function toDays(iso) {
  const [y, m, d] = iso.split('-').map(Number);
  return Math.floor(Date.UTC(y, m - 1, d) / 86400000);
}

/** Convert a whole-day number back to a "YYYY-MM-DD" string. */
export function fromDays(n) {
  return new Date(n * 86400000).toISOString().slice(0, 10);
}

/** Weekday index for a date, Monday=0 through Sunday=6 (matches the UI order). */
export function weekdayOf(iso) {
  return (new Date(toDays(iso) * 86400000).getUTCDay() + 6) % 7;
}

/**
 * Month and weekday names, fixed rather than taken from the browser locale.
 *
 * Locale formatting would vary by device (en-GB renders September as "Sept",
 * four characters where every other month gives three) which both breaks the
 * column alignment in the text table and means two people copying the same
 * roster could get different text. A roster is shared output, so it is pinned.
 */
const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
                'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
const WEEKDAY_NAMES = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

/** Short weekday name for a date, e.g. "Sat". */
export function weekdayName(iso) {
  return WEEKDAY_NAMES[weekdayOf(iso)];
}

/** Format a date for display, e.g. "Sat 06 Jun 2026". */
export function formatDate(iso, { withWeekday = true } = {}) {
  const [y, m, d] = iso.split('-').map(Number);
  const stamp = `${String(d).padStart(2, '0')} ${MONTHS[m - 1]} ${y}`;
  return withWeekday ? `${weekdayName(iso)} ${stamp}` : stamp;
}

/**
 * Every date from start to end inclusive whose weekday is rostered.
 * `weekdays` is a Set of 0-6 values; an empty set yields no days at all.
 */
export function dateRange(start, end, weekdays) {
  let a = toDays(start);
  let b = toDays(end);
  if (b < a) [a, b] = [b, a];

  const out = [];
  for (let n = a; n <= b; n++) {
    const iso = fromDays(n);
    if (weekdays.has(weekdayOf(iso))) out.push(iso);
  }
  return out;
}

/** Expand unordered clash pairs into a person-id -> Set(incompatible ids) map. */
function clashMap(clashes) {
  const map = new Map();
  for (const [a, b] of clashes) {
    if (!map.has(a)) map.set(a, new Set());
    if (!map.has(b)) map.set(b, new Set());
    map.get(a).add(b);
    map.get(b).add(a);
  }
  return map;
}

/**
 * True if a person's standing availability allows working `iso`.
 * Mirrors Person.can_work() from the Python original.
 */
export function canWork(person, iso) {
  if (!person.active) return false;
  if (!person.availableWeekdays.includes(weekdayOf(iso))) return false;
  const n = toDays(iso);
  return !(person.blackouts || []).some(
    (b) => toDays(b.start) <= n && n <= toDays(b.end)
  );
}

/**
 * Choose the fairest eligible person for one vacancy, or null if there is none.
 */
function pick(day, people, usedToday, placedIds, incompatible, worked, lastWorked) {
  const eligible = people.filter(
    (p) =>
      !usedToday.has(p.id) &&                                  // not already on today
      canWork(p, day) &&                                       // availability + blackouts
      ![...(incompatible.get(p.id) || [])].some((id) => placedIds.has(id)) && // no clash
      (p.maxShifts == null || worked.get(p.id) < p.maxShifts)
  );
  if (eligible.length === 0) return null;

  // Fairness ranking - see the module comment. Comparing `name` last makes the
  // ordering total, so the result is reproducible run to run.
  eligible.sort((x, y) => {
    const wx = worked.get(x.id), wy = worked.get(y.id);
    if (wx !== wy) return wx - wy;
    const lx = lastWorked.get(x.id), ly = lastWorked.get(y.id);
    if (lx !== ly) return lx - ly;
    return x.name.toLowerCase().localeCompare(y.name.toLowerCase());
  });
  return eligible[0];
}

/**
 * Allocate staff across the date range and return the finished roster.
 *
 * Anything that could not be satisfied - an unfillable shift, a person with no
 * availability at all - is pushed onto `roster.notes` instead of being silently
 * dropped, so the UI can surface it next to the result.
 */
export function buildRoster({ orgId, name, people, shifts, clashes, start, end, weekdays, dates }) {
  // `dates` is the canonical input: the exact days to roster, as chosen on the
  // calendar. start/end/weekdays remain supported as a shorthand that derives
  // the same list, which is what the test suite and older saved rosters use.
  // Presence, not length, picks the branch - an empty `dates` means "nothing
  // selected" and must report that, not silently fall back to a date range
  // whose bounds were never supplied.
  const days = Array.isArray(dates)
    ? [...new Set(dates)].sort()
    : dateRange(start, end, weekdays);
  const roster = {
    orgId,
    name,
    startDate: days.length ? days[0] : start,
    endDate: days.length ? days[days.length - 1] : end,
    // The exact dates rostered, kept on the roster so a saved one can be
    // redrawn faithfully later. Without it a day where nothing could be filled
    // would simply vanish from the saved view, hiding the very gap that most
    // needs looking at.
    days,
    assignments: [],
    notes: [],
  };

  // Guard the degenerate inputs up front so the main loop can assume it has
  // something to work with.
  const roll = people.filter((p) => p.active);
  if (days.length === 0) {
    roster.notes.push(dates
      ? 'No dates picked on the calendar.'
      : 'No rostered days fall in the selected date range.');
    return roster;
  }
  if (shifts.length === 0) {
    roster.notes.push('This organisation has no shifts set up.');
    return roster;
  }
  if (roll.length === 0) {
    roster.notes.push('No active people to roster.');
    return roster;
  }

  const incompatible = clashMap(clashes);
  const worked = new Map(roll.map((p) => [p.id, 0]));
  const lastWorked = new Map(roll.map((p) => [p.id, NEVER]));

  for (const day of days) {
    // One shift per person per day: whoever is already on today is out of the
    // pool for the day's remaining shifts.
    const usedToday = new Set();

    for (const shift of shifts) {
      const placedIds = new Set();
      let placed = 0;

      for (let i = 0; i < Math.max(0, shift.headcount); i++) {
        const who = pick(day, roll, usedToday, placedIds, incompatible, worked, lastWorked);
        if (!who) break;
        placed++;
        placedIds.add(who.id);
        usedToday.add(who.id);
        worked.set(who.id, worked.get(who.id) + 1);
        lastWorked.set(who.id, toDays(day));
        roster.assignments.push({ date: day, shiftId: shift.id, personId: who.id });
      }

      // Report the gap rather than pretending the shift is covered.
      if (placed < shift.headcount) {
        roster.notes.push(
          `${formatDate(day)} - ${shift.name}: only ${placed} of ${shift.headcount} filled ` +
          `(nobody else was available).`
        );
      }
    }
  }

  // Flag people the roster never used, which is nearly always a data problem
  // (no availability ticked, or a blackout covering the whole period).
  for (const p of roll) {
    if (worked.get(p.id) === 0) {
      const reason = p.availableWeekdays.length === 0
        ? 'no days ticked in their availability'
        : 'unavailable on every rostered day';
      roster.notes.push(`${p.name} was not rostered at all - ${reason}.`);
    }
  }

  return roster;
}

/** Shifts worked per person id, for the fairness summary. */
export function counts(roster) {
  const out = new Map();
  for (const a of roster.assignments) {
    out.set(a.personId, (out.get(a.personId) || 0) + 1);
  }
  return out;
}

/**
 * Render a roster as an aligned monospace table.
 *
 * Used for the Copy and Export actions - it pastes cleanly into a message or a
 * printout. The on-screen view uses a day-by-day layout instead, because a wide
 * table is the wrong shape for a phone.
 */
export function formatTable(roster, people, shifts) {
  const names = new Map(people.map((p) => [p.id, p.name]));

  // Index assignments by date+shift so each cell is a single lookup.
  const cells = new Map();
  for (const a of roster.assignments) {
    const key = `${a.date}|${a.shiftId}`;
    if (!cells.has(key)) cells.set(key, []);
    cells.get(key).push(names.get(a.personId) || `#${a.personId}`);
  }

  const days = [...new Set(roster.assignments.map((a) => a.date))].sort();
  const headers = ['Date', 'Day', ...shifts.map((s) => shiftLabel(s))];
  const rows = days.map((d) => [
    formatDate(d, { withWeekday: false }),
    weekdayName(d),
    ...shifts.map((s) => (cells.get(`${d}|${s.id}`) || []).sort().join(', ') || '-'),
  ]);

  // Column widths are the widest cell in each column, header included.
  const widths = headers.map((h, i) =>
    Math.max(h.length, ...rows.map((r) => r[i].length), 0)
  );
  const line = (cols) =>
    cols.map((c, i) => c.padEnd(widths[i])).join('  ').replace(/\s+$/, '');

  const out = [line(headers), widths.map((w) => '-'.repeat(w)).join('  ')];
  for (const r of rows) out.push(line(r));

  // Fairness summary: how many shifts each person actually ended up with.
  const tally = [...counts(roster).entries()].sort(
    (a, b) => b[1] - a[1] || (names.get(a[0]) || '').localeCompare(names.get(b[0]) || '')
  );
  if (tally.length) {
    out.push('', 'Summary');
    for (const [pid, n] of tally) {
      out.push(`  ${names.get(pid) || `#${pid}`}: ${n} shift${n === 1 ? '' : 's'}`);
    }
  }

  if (roster.notes.length) {
    out.push('', 'Notes');
    for (const n of roster.notes) out.push(`  - ${n}`);
  }

  return out.join('\n');
}

/** Display label for a shift, e.g. "Early (06:00-14:00)". */
export function shiftLabel(shift) {
  if (shift.start && shift.end) return `${shift.name} (${shift.start}-${shift.end})`;
  return shift.name;
}

File: legacy-desktop/core/scheduler.py
"""Deterministic roster allocation.

This is the part that replaces the Guardian bot's "send the constraints to an
LLM and print whatever comes back". Given the same inputs this module always
produces the same roster, every constraint it is given is actually enforced,
and anything it *cannot* satisfy is reported as an explicit note rather than
quietly fudged.

The algorithm is a greedy fill with a fairness ranking. Days are processed in
order; within a day, shifts are filled in the organization's own shift order.
For each vacancy the eligible pool is ranked by

    1. fewest shifts worked so far in this roster   (spread the load)
    2. longest since last worked                    (spread the days out)
    3. name                                         (a stable, arbitrary tiebreak)

and the top candidates are taken. Greedy rather than an exhaustive search
because rosters are small, the result must be explainable to the person reading
it ("Alice got Tuesday because she'd worked the fewest shifts"), and a solver
that silently reshuffles everything on a one-person change is worse to live
with than one that does the obvious thing.
"""
from __future__ import annotations

from datetime import date, timedelta

from core.models import Assignment, Person, Roster, ShiftType

#: Sentinel "has never worked" date used by the recency tiebreak. Any real date
#: sorts after it, so people who have not worked yet are always preferred.
_NEVER = date.min


def date_range(start: date, end: date, weekdays: set[int]) -> list[date]:
    """Every date from ``start`` to ``end`` inclusive whose weekday is rostered.

    ``weekdays`` holds ``date.weekday()`` values (0=Monday .. 6=Sunday); an
    empty set means no days at all, which the caller should treat as an error.
    """
    if end < start:
        start, end = end, start
    out: list[date] = []
    day = start
    while day <= end:
        if day.weekday() in weekdays:
            out.append(day)
        day += timedelta(days=1)
    return out


def _clash_map(clashes: list[tuple[int, int]]) -> dict[int, set[int]]:
    """Expand unordered clash pairs into a person-id -> incompatible-ids map."""
    out: dict[int, set[int]] = {}
    for a, b in clashes:
        out.setdefault(a, set()).add(b)
        out.setdefault(b, set()).add(a)
    return out


def build_roster(
    org_id: int,
    name: str,
    people: list[Person],
    shifts: list[ShiftType],
    clashes: list[tuple[int, int]],
    start: date,
    end: date,
    weekdays: set[int],
) -> Roster:
    """Allocate staff across the date range and return the finished roster.

    Anything that could not be satisfied - an unfillable shift, a person with
    no availability at all - is appended to ``roster.notes`` instead of being
    silently dropped, so the GUI can surface it next to the table.
    """
    days = date_range(start, end, weekdays)
    roster = Roster(org_id=org_id, name=name,
                    start_date=days[0] if days else start,
                    end_date=days[-1] if days else end)

    # Guard the degenerate inputs up front so the main loop can assume it has
    # something to work with.
    roster_people = [p for p in people if p.active]
    if not days:
        roster.notes.append("No rostered days fall in the selected date range.")
        return roster
    if not shifts:
        roster.notes.append("This organization has no shift types defined.")
        return roster
    if not roster_people:
        roster.notes.append("No active staff to roster.")
        return roster

    incompatible = _clash_map(clashes)

    # Running fairness state, keyed by person id.
    worked: dict[int, int] = {p.id: 0 for p in roster_people}
    last_worked: dict[int, date] = {p.id: _NEVER for p in roster_people}

    for day in days:
        # One shift per person per day: whoever is already on today is out of
        # the pool for the day's remaining shifts.
        used_today: set[int] = set()

        for shift in shifts:
            placed: list[Person] = []

            for _ in range(max(0, shift.headcount)):
                candidate = _pick(shift, day, roster_people, used_today, placed,
                                  incompatible, worked, last_worked)
                if candidate is None:
                    break
                placed.append(candidate)
                used_today.add(candidate.id)
                worked[candidate.id] += 1
                last_worked[candidate.id] = day
                roster.assignments.append(
                    Assignment(work_date=day, shift_type_id=shift.id,
                               person_id=candidate.id))

            # Report the gap rather than pretending the shift is covered.
            if len(placed) < shift.headcount:
                roster.notes.append(
                    f"{day.strftime('%a %d %b %Y')} - {shift.name}: "
                    f"only {len(placed)} of {shift.headcount} filled "
                    f"(nobody else was available)."
                )

    # Flag staff the roster never used, which is nearly always a data problem
    # (no availability ticked, or a blackout covering the whole period).
    for p in roster_people:
        if worked[p.id] == 0:
            reason = ("no weekdays ticked in their availability"
                      if not p.available_weekdays
                      else "unavailable on every rostered day")
            roster.notes.append(f"{p.name} was not rostered at all - {reason}.")

    return roster


def _pick(
    shift: ShiftType,
    day: date,
    people: list[Person],
    used_today: set[int],
    placed: list[Person],
    incompatible: dict[int, set[int]],
    worked: dict[int, int],
    last_worked: dict[int, date],
) -> Person | None:
    """Choose the fairest eligible person for one vacancy, or None if there is none."""
    placed_ids = {p.id for p in placed}

    eligible = [
        p for p in people
        if p.id not in used_today                       # not already on today
        and p.can_work(day)                             # availability + blackouts
        and not (incompatible.get(p.id, set()) & placed_ids)   # no clash on this shift
        and (p.max_shifts is None or worked[p.id] < p.max_shifts)
    ]
    if not eligible:
        return None

    # Fairness ranking - see the module docstring. `name` last makes the whole
    # ordering total, so the result is reproducible run to run.
    eligible.sort(key=lambda p: (worked[p.id], last_worked[p.id], p.name.lower()))
    return eligible[0]


def format_table(roster: Roster, people: list[Person],
                 shifts: list[ShiftType]) -> str:
    """Render a roster as the aligned monospace table shown in the app and exports.

    Columns are Date | Day | one per shift type, followed by a per-person shift
    summary and any notes the scheduler raised.
    """
    names = {p.id: p.name for p in people}
    shift_list = list(shifts)

    # Index assignments by (date, shift) so each cell is a single lookup.
    cells: dict[tuple[date, int], list[str]] = {}
    for a in roster.assignments:
        cells.setdefault((a.work_date, a.shift_type_id), []).append(
            names.get(a.person_id, f"#{a.person_id}"))

    days = sorted({a.work_date for a in roster.assignments}) or \
        date_range(roster.start_date, roster.end_date, set(range(7)))

    headers = ["Date", "Day"] + [s.label for s in shift_list]
    rows = [
        [d.strftime("%d %b %Y"), d.strftime("%a")] +
        [", ".join(sorted(cells.get((d, s.id), []))) or "-" for s in shift_list]
        for d in days
    ]

    # Column widths are the widest cell in each column, header included.
    widths = [max(len(headers[i]), *(len(r[i]) for r in rows)) if rows
              else len(headers[i]) for i in range(len(headers))]

    def line(cols: list[str]) -> str:
        """Pad one row's cells out to the computed column widths."""
        return "  ".join(c.ljust(widths[i]) for i, c in enumerate(cols)).rstrip()

    out = [line(headers), "  ".join("-" * w for w in widths)]
    out += [line(r) for r in rows]

    # Fairness summary: how many shifts each person actually ended up with.
    counts = roster.counts()
    if counts:
        out += ["", "Summary"]
        for pid, n in sorted(counts.items(),
                             key=lambda kv: (-kv[1], names.get(kv[0], ""))):
            out.append(f"  {names.get(pid, f'#{pid}')}: {n} shift"
                       f"{'s' if n != 1 else ''}")

    if roster.notes:
        out += ["", "Notes"]
        out += [f"  - {n}" for n in roster.notes]

    return "\n".join(out)
